src/client.ts

```diff
@@ -852,6 +938,209 @@ export class StreamClient<T extends QueryValue = any> {
+    };
+
+    // Set the page size if it is available
+    if (this.#clientConfiguration.page_size) {
+      req.data.page_size = this.#clientConfiguration.page_size;
+    }
+
+    // If we have a cursor, use that. Otherwise, use the start_ts if available.
+    // When the config is validated, if both are set, an error is thrown.
+    if (this.#lastCursor) {
+      req.data.cursor = this.#lastCursor;
+    } else if (this.#clientConfiguration.start_ts) {
+      req.data.start_ts = this.#clientConfiguration.start_ts;
+    }
+
+    return req;
+  }
+
+  async *[Symbol.asyncIterator](): AsyncGenerator<ChangeFeedPage<T>> {
+    while (!this.#isDone) {
+      yield await this.nextPage();
+    }
+  }
+
+  /**
+   * Fetches the next page of the change feed. If there are no more pages to
+   * fetch, this method will throw a {@link ClientError}.
+   */
+  async nextPage(): Promise<ChangeFeedPage<T>> {
+    if (this.#isDone) {
+      throw new ClientError("The change feed has no more pages to fetch.");
+    }
+
+    const { httpClient } = this.#clientConfiguration;
+
+    const request = await this.#nextPageHttpRequest();
+    const response = await withRetries(() => httpClient.request(request), {
+      maxAttempts: this.#clientConfiguration.max_attempts,
+      maxBackoff: this.#clientConfiguration.max_backoff,
+      shouldRetry: (error) => error instanceof ThrottlingError,
+    });
+
+    let body: ChangeFeedSuccess<T> | ChangeFeedError;
+
+    try {
+      body = TaggedTypeFormat.decode(response.body, {
+        long_type: this.#clientConfiguration.long_type,
+      });
+    } catch (error: unknown) {
+      throw new ProtocolError({
+        message: `Error parsing response as JSON: ${error}`,
+        httpStatus: response.status,
+      });
+    }
+
+    if (isQueryFailure(body)) {
+      throw getServiceError(body, response.status);
+    }
+
+    const page = new ChangeFeedPage<T>(body);
+    this.#lastCursor = page.cursor;
+    this.#isDone = !page.hasNext;
+
+    return page;
+  }
+
+  /**
+   * Returns an async generator that yields the events of the change feed
+   * directly.
+   *
+   * @example
+   * ```javascript
+   *  const changeFeed = client.changeFeed(fql`MyCollection.all().toStream()`)
+   *
+   *  for await (const user of changeFeed.flatten()) {
+   *    // do something with each event
+   *  }
+   * ```
+   */
+  async *flatten(): AsyncGenerator<StreamEventData<T>> {
+    for await (const page of this) {
+      for (const event of page.events) {
+        yield event;
+      }
+    }
+  }
+
+  async #resolveStreamToken(
+    fn: () => Promise<StreamToken>,
+  ): Promise<StreamToken> {
+    return await fn().then((maybeStreamToken) => {
+      if (!(maybeStreamToken instanceof StreamToken)) {
+        throw new ClientError(
+          `Error requesting a stream token. Expected a StreamToken as the query result, but received ${typeof maybeStreamToken}. Your query must return the result of '<Set>.toStream' or '<Set>.changesOn')\n` +
```

Comment:
`typeof` is going to be the run-time type, right? it may more useful to just safely stringify (`util.inspect`?) it instead; seems like it'll be "Object" in almost all misconfigurations.

Reply:
Good point about `util.inspect`. Our domain types do not have toJSON overrides, so for example JSON.stringify(some_time_stub)` will result in something like `{ isoString: "2024-10-10T00:00:00Z" }`. So, at least `util.inspect` would show `TimeStub { isoString: "2024-10-10T00:00:00Z" }`.

How about using `${maybeStreamToken.constructor.name}` here and then `util.inspect` at the end of the message?

Reply:
i'm ok with that.

util.inspect has the added bonus that self-referential objects don't break the runtime.

Reply:
I'll do this as a follow-up PR as it's going to change in two places (and one is unrelated to this PR). This is already a chonky PR.